snapcraft/internal/lxd.py

```diff
@@ -218,6 +218,25 @@ def _inject_snap(self, name):
         id = json['result']['id']
         # Lookup confinement to know if we need to --classic when installing
         is_classic = json['result']['confinement'] == 'classic'
+
+        # If the server has a different arch we can't inject local snaps
+        if (self._project_options.target_arch
+                and self._project_options.target_arch != self._server_arch):
+            logger.info('Different arch {} != {}'.format(
+                self._project_options.deb_arch, self._server_arch))
+            logger.info('Installing {}'.format(name))
+            # Install if needed
+            cmd = ['snap', 'install', name]
+            if is_classic:
+                cmd.append('--classic')
+            cmd.append('--channel')
+            cmd.append(json['result']['channel'])
+            self._container_run(cmd)
+            # Switch channel if needed
+            cmd[1] = 'refresh'
+            self._container_run(cmd)
+            return
```

Comment:
I think it might be clearer if you split wrap this into a function.

Reply:
Good idea actually. I refactored it now, and the file-based install also uses it.